File: 6. ZigZag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        store = []
        for i in range(numRows):
            store.append('')
        flag = 1
        pos = 1
        for i, r in enumerate(s):
            if pos == 1:
                flag = 1
            elif pos == numRows:
                flag = -1
            store[pos-1] += r
            pos += flag
        return ''.join(store)

File: 6. ZigZag Conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }
        List<StringBuilder> rows = new ArrayList<>();
        for (int i =0; i<numRows; i++) {
            rows.add(new StringBuilder());
        }
        int currRow = 0;
        boolean goingDown = false;
        for (char c : s.toCharArray()) {
            rows.get(currRow).append(c);
            if (currRow == 0 || currRow == numRows-1) {
                goingDown = !goingDown;
            }
            currRow += goingDown ? 1 : -1;
        }
        StringBuilder sb = new StringBuilder();
        for (StringBuilder row : rows) {
            sb.append(row);
        }
        return sb.toString();
    }
}